### PR TITLE
May 2024

### PR DESCRIPTION
**Pull request by ballsteve**
This PR implements the remaining features required to make Xrust functionally equivalent to XPath 1.0 and XSLT 1.0.

The following features have been implemented:

* modal template processing
* XML Namespace node support
* Sorting for xsl:apply-templates, xsl:for-each, and xsl:for-each-group
* document() function
* format-number() function
* xsl:number element
* xsl:attribute-sets and xsl:use-attribute-sets
* XPath abbreviated syntax

### CODE REVIEW

File: examples/issue-30.rs [hunk 1]
```diff
@@ -60,20 +64,24 @@ fn main() {
...
     // Parse the XML into a RNode
-    // content[0] is the root node
-    let root = Document::try_from((srcxml, None, None))
-        .expect("unable to parse XML")
-        .content[0]
-        .clone();
-    let mut doc = NodeBuilder::new(NodeType::Document).build();
-    doc.push(root).expect("unable to append root node");
+    let root = Rc::new(SmiteNode::new());
+    xmlparse(root.clone(), srcxml.as_str(), None, None).expect("unable to parse XML");
```
Comment (makigumo) [inline examples/issue-30.rs:68]:
```suggestion
    xmlparse(root.clone(), srcxml.as_str(), None).expect("unable to parse XML");
```

following the change in commit [5077833bdddd2b1a4780893fc498d4808a70806b](https://github.com/ballsteve/xrust/commit/5077833bdddd2b1a4780893fc498d4808a70806b#diff-0eabaaeb1e42c9e6de94c303f7fd31aeba580aa78bc0c0e66f6ac97506e47b46L31-R31)

File: examples/ixml.rs [hunk 2]
```diff
@@ -119,12 +116,8 @@ fn main() {
...
     };
 
-    let mut style = NodeBuilder::new(NodeType::Document).build();
-    let style_doc =
-        Document::try_from((stylexml.trim(), None, None)).expect("failed to parse XSL stylesheet");
-    style
-        .push(style_doc.content[0].clone())
-        .expect("unable to append style nodes");
+    let style = Rc::new(SmiteNode::new());
+    parse(style.clone(), stylexml.trim(), None, None).expect("failed to parse XSL stylesheet");
```
Comment (makigumo) [inline examples/ixml.rs:120]:
```suggestion
    parse(style.clone(), stylexml.trim(), None).expect("failed to parse XSL stylesheet");
```
same as above

File: src/parser/xpath/mod.rs [hunk 3]
```diff
@@ -16,24 +16,28 @@ To evaluate the transformation we need a Context with a source document as its c
...
 use xrust::parser::xpath::parse;
-use xrust::transform::context::{Context, ContextBuilder, StaticContext};
-# type F = Box<dyn FnMut(&str) -> Result<(), Error>>;
+use xrust::transform::context::{Context, ContextBuilder, StaticContext, StaticContextBuilder};
 
 let t = parse("/child::A/child::B/child::C")
     .expect("unable to parse XPath expression");
 
 let source = Rc::new(SmiteNode::new());
 xmlparse(source.clone(), "<A><B><C/></B><B><C/></B></A>", None, None)
```
Comment (makigumo) [inline src/parser/xpath/mod.rs:30]:
```suggestion
xmlparse(source.clone(), "<A><B><C/></B><B><C/></B></A>", None)
```
same as above

**Comment by Devasta** [comment]
Hi @makigumo,

Many thanks for your reviews of the code, those suggestions will be merged. 😄

**Comment by Devasta** [comment]
Hi @ballsteve ,

   xpath_for_1
    xpath_for_2
    xpath_if_1
    xpath_if_2
    xpath_let_1
    xpath_let_2


Six failing tests for me on Windows, error is:

    ---- xpath_for_1 stdout ----
    thread 'xpath_for_1' panicked at tests\smite-xpath.rs:431:10:
    test failed: Unrecognised extra characters: "Unrecognised extra characters: " $x in ('a', 'b', 'c') return ($x, $x)""

Checking now.

**Comment by ballsteve** [comment]
Looks like I jumped the gun a bit... I must have gotten too excited about getting the last of those features finished. Those errors will be due to the changes required for abbreviated XPath syntax, so I'll take a look at that.

**Comment by ballsteve** [comment]
Path parser bug now fixed in may-2024 branch

**Comment by Devasta** [comment]
Looks good! Merging to Dev now.

**Pull request merged**